.clang-format

```diff
@@ -1,10 +1,100 @@
+Language: Cpp
+Standard: c++20
+ColumnLimit: 100
+
 WhitespaceSensitiveMacros:
   # clang format doesn't understand TypeScript, so make sure it doesn't mangle
   # overrides and additional definitions
   - JSG_TS_OVERRIDE
   - JSG_TS_DEFINE
   - JSG_STRUCT_TS_OVERRIDE
   - JSG_STRUCT_TS_DEFINE
-PointerAlignment: Left
-ColumnLimit: 100
 AllowShortFunctionsOnASingleLine: Empty
+
+# We should have "true" here but adding an include header would result in too many changed lines.
+# Once we turn this on we should use IncludeCategories so that things are consistently sorted.
+# For example, to protect against brittle headers (a header not including everything its using
+# because existing users happened to include a dependency for it), one could use the following
+# order:
+#   local folder files
+#   project files
+#   3p dependency includes
+#   standard language headers (put kj/ here?)
+#   system headers
+# While this would be ideal it's also important to note that this isn't the (non-documented) style
+# that KJ uses, so it may be worth documenting the style & making it consistent.
+SortIncludes: false
```

Comment:
This might be a good time to define IncludeCategories and turn this flag on.
https://clang.llvm.org/docs/ClangFormatStyleOptions.html#includecategories

Reply:
I disagree – we don't have this set up internally either and as suggested by the comments getting include sorting right is non-trivial. IMHO formatting code correctly is more important than formatting includes correctly and getting includes right is likely to slow landing this down

Reply:
Why is getting include sorting right non-trivial though?
What's wrong with something like:
```
SortIncludes: true
IncludeBlocks: Regroup
IncludeCategories:
  # c++ system headers
  - Regex: <[a-zA-Z0-9_]+>
    Priority: 5
  # kj/capnp headers
  - Regex: <(kj|capnp)/.+>
    Priority: 4
  # 3rd party headers
  - Regex: <.+>
    Priority: 3
  # workerd headers
  - Regex: <workerd/.+>
    Priority: 2
  # local headers
  - Regex: '".*"'
    Priority: 1
```

Reply:
The categories here look reasonable, but if b.c++ includes a.h and b.h, does this ensure that b.h is listed first to check that b.h can be included without depending on other headers? We can try checking includes, my concern is that this will delay landing this PR if there's any complications with include changes, it's already pretty expansive.

Reply:
So for any <filename>.c++ the first include will always be "<filename>.h", with the above config.
And any header file should always be include-able without first including other headers, if that's not true then said head file should be fixed.

Reply:
Here's what clang's manual says about this:

> The main header for a source file automatically gets category 0. so that it is generally kept at the beginning of the #includes (https://llvm.org/docs/CodingStandards.html#include-style).

Reply:
To summarise here, I think adding header include sorting is a huge +.
But I do understand Felix's worries.
I will defer to @jp4a50 whether we should add it to this PR or a later one. Either way the above configuration should be fitting and can be used at a later date.

Reply:
To enable sortincludes with the above config and have workerd compile and pass all tests the following changes need to be made:
src/workerd/jsg/jsg.h: Add clang format off around the includes in the end.

// clang-format off
// These two includes are needed for the JSG type glue macros to work.
#include "buffersource.h"
#include "modules.h"
#include "resource.h"
#include "dom-exception.h"
#include "struct.h"
#include "promise.h"
#include "function.h"
#include "iterator.h"
#include "jsvalue.h"
// clang-format on


src/workerd/api/memory-cache.h: remove the include to function.h
delete `#include <workerd/jsg/function.h>`